add fnn-vae (#97)
rstudio/ai-blog: _posts/2020-07-31-fnn-vae-for-noisy-timeseries/fnn_vae_for_noisy_timeseries.Rmd

with a whole new smoothness of predictions; but FNN-LSTM turns up much smoother forecasts as well.

"Smooth, fitting the system..." -- by now you may be wondering, when are we going to come up with more quantitative
assertions? If quantitative implies "mean squared error" (MSE), and if MSE is taken to be some divergence between forecasts
and the true target from the test set, the answer is that this MSE doesn't differ much between any of the four architectures.
Put differently, it is mostly a function of noise level.

However, we could argue that what we're really interested in is how well a model forecasts the underlying process. And there,
we see differences.

In the following plot, we contrast MSEs obtained for the four model types (grey: VAE; orange: LSTM; dark blue: FNN-VAE; green:
FNN-LSTM). The rows reflect noise levels (1, 1.5, 2, 2.5); the columns represent MSE in relation to the noisy("real") target
(left) on the one hand, and in relation to the underlying system on the other (right). For better visibility of the effect,
*MSEs have been normalized as fractions of the maximum MSE in a category*.

> So, if we want to predict *signal plus noise* (left), it doesn't really matter whether we use FNN or not. But if we want to

Reviewer: I wouldn't say it doesn't matter as from the plot it seems that FNN regularised versions are consistently slightly better for higher noise values.

Author: you're right! I really wanted to be as cautious as possible ... But if you say so, i'll rephrase that a bit :-)

predict the signal only (right), with increasing noise in the data FNN loss becomes increasingly effective. This effect is far
stronger for VAE vs. FNN-VAE than for LSTM vs. FNN-LSTM: The distance between the grey line (VAE) and the dark blue one
(FNN-VAE) becomes larger and larger as we add more noise.

> ```{r, eval=TRUE, echo=FALSE, layout="l-page", fig.cap = "Normalized MSEs obtained for the four model types (grey: VAE; orange: LSTM; dark blue: FNN-VAE; green: FNN-LSTM). Rows are noise levels (1, 1.5, 2, 2.5); columns are MSE as related to the real target (left) and the underlying system (right)."}

Reviewer: It took me some time to clearly understand the plot, I had to come back and forth to the text multiple times to remember what's in each column/row. Also there's no description of what's the `x` axis.. I assumed that it's the prediction step.

Reviewer 2: Oh +1 on this—I usually proof in Rmd, so I hadn't looked at the preview, but I'd definitely add a legend (check out [patchwork](https://github.com/thomasp85/patchwork) or [cowplot](https://github.com/wilkelab/cowplot), maybe, if you want finer-tuned control). Let me know if you need any help!

Author: i'm using cowplot ... i didn't use any legends anywhere, to save space, as anyway i'm always having the text below the figures....

Author: changed the text to say what's on the x axis

---
title: "FNN-VAE for noisy time series forecasting"
description: > 

>  In this last part of a mini-series on forecasting with false nearest neighbors (FNN) loss, we replace the LSTM autoencoder from the previous post by a convolutional VAE, resulting in equivalent prediction performance but significantly lower training time. In addition, we find that FNN regularization is of great help when an underlying deterministic process is obscured by substantial noise.

Author: thanks, committing :-) And no, it's not a series in that way (that formally, I mean)

  
author:
  - name: Sigrid Keydana
    affiliation: RStudio
    affiliation_url: https://www.rstudio.com/
slug: keydanafnnvae
bibliography: bibliography.bib
date: 07-31-2020
categories:
  - R
  - TensorFlow/Keras
  - Time Series
  - Unsupervised Learning
output:
  distill::distill_article:
    self_contained: false
preview: images/kb.jpg
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

This post did not quite end up the way I'd imagined. A quick follow-up on the recent [Time series prediction with

> FNN-LSTM](https://blogs.rstudio.com/ai/posts/2020-07-20-fnn-lstm/), it should have shown how *noisy time series* (so common in

Author: the former ... will change to "was supposed to", perhaps that makes it clearer

practice) could profit from a change in architecture: Instead of FNN-LSTM, an LSTM autoencoder regularized by false nearest
neighbors (FNN) loss, use FNN-VAE, a variational autoencoder constrained by the same. However, FNN-VAE did not seem to handle
noise better than FNN-LSTM. No plot, no post, then?

> On the other hand -- this is not a scientific study, with hypothesis and experimental setup all preregistered; all that really

Author: hm, yeah... I think that in this special case, the rhythm really needs the semicolon (or perhaps what I mean is intonation? voice going down, not up...

> probably more a visual thing than anything else.

in that case... yeah, for me this is auditory, this one for sure...

matters is if there's something useful to report. And it looks like there is.

Firstly, FNN-VAE, while on par, performance-wise, with FNN-LSTM, it is far superior in that other meaning of "performance":
Training goes a *lot* faster for FNN-VAE.

And secondly, while we don't see much difference between FNN-LSTM and FNN-VAE, we *do* see a clear impact of using FNN loss:

> Especially for the VAE architecture, but for the LSTM one as well, adding in FNN loss strongly reduces mean squared error with

Author: thanks for the link, I'll take a look! I do tend to think by now that really there be something of a usage difference between languages here: (hehe)

> I would break it into two sentences, 

For me, phrases separated by colons nearly always ARE separate sentences! Will change; and put more attention to it in the future (when reading English prose, I mean)

respect to the underlying (denoised) series. This is of particular interest with VAE, as it comes with a regularizer
out-of-the-box -- namely, Kullback-Leibler (KL) divergence.

> Of course, we don't claim that similar results will always be obtained on other noisy series; and neither did we tune any of

Author: will do!

the models "to death". For, what could be the intent of such a post but to show our readers interesting (and promising) ideas
to pursue in their own experimentation?

## The context

This post is the third in a mini-series.

In [Deep attractors: Where deep learning meets chaos](https://blogs.rstudio.com/ai/posts/2020-06-24-deep-attractors/), we

> explained, with a substantial detour into chaos theory, the idea of FNN loss, introduced in [@gilpin2020deep]. Please consult

Author: I get the point, but don't see how to make that work exactly ...will try to keep in mind for other occasions though

that first post for theoretical background and intuitions behind the technique.

The subsequent post, [Time series prediction with FNN-LSTM](https://blogs.rstudio.com/ai/posts/2020-07-20-fnn-lstm/), showed

> how to use it for prediction (as opposed to reconstructing an attractor). "It", the technique, meant augmenting an LSTM

Author: makes sense!! was never happy with that one either

> autoencoder with FNN loss. The results were stunning: In multi-step prediction (12-120 steps, with that number varying by

Author: grammar-nerd stuff always welcome :-)

dataset), the short-term forecasts were drastically improved by adding in FNN regularization. See that second post for
experimental setup and results on four very different, non-synthetic datasets.

Today, we show how to replace the LSTM autoencoder by a -- convolutional -- VAE. In light of the experimentation results,

> already hinted at above, it is completely thinkable that the "variational" part is not even so important here -- that a

Author: yep! non-native speaker's thing here i guess

convolutional autoencoder with just MSE loss would have performed just as well on those data. In fact, to find out, it's
enough to just remove the call to `reparameterize()` and multiply the KL component of the loss by 0. (We leave this to the
interested reader, to keep the post at reasonable length.)

> One last contextual thing, in case you haven't read the two previous posts and would like to jump in here directly. We're

Reviewer 2: ```suggestion
One last piece of context, in case you haven't read the two previous posts and would like to jump in here directly. We're
```

doing time series forecasting; so why this talk of autoencoders? Shouldn't we just be comparing an LSTM (or some other type of
RNN, for that matter) to a convnet? In fact, the necessity of a latent representation is due to the very idea of FNN: The
latent code is supposed to reflect the true attractor of a dynamical system. That is, if the attractor of the underlying

> system is roughly two-dimensional, we hope to find that just two of the latent variables have considerable variance. (This

Reviewer 2: Really nicely done! That's one heck of a thing to describe, but the sentence flows very nicely.

    self$act1 <- layer_activation_leaky_relu()
    self$batchnorm1 <- layer_batch_normalization()
    self$conv2 <- layer_conv_1d(kernel_size = 7,
                                filters = 32,
                                strides = 2)
    self$act2 <- layer_activation_leaky_relu()
    self$batchnorm2 <- layer_batch_normalization()
    self$conv3 <- layer_conv_1d(kernel_size = 9,
                                filters = 64,
                                strides = 2)
    self$act3 <- layer_activation_leaky_relu()
    self$batchnorm3 <- layer_batch_normalization()
    self$conv4 <- layer_conv_1d(
      kernel_size = 9,
      filters = n_latent,
      strides = 2,
      activation = "linear" 
    )
    self$batchnorm4 <- layer_batch_normalization()
    self$flat <- layer_flatten()
    
    function (x, mask = NULL) {
      x %>%
        self$conv1() %>%
        self$act1() %>%
        self$batchnorm1() %>%
        self$conv2() %>%
        self$act2() %>%
        self$batchnorm2() %>%
        self$conv3() %>%
        self$act3() %>%
        self$batchnorm3() %>%
        self$conv4() %>%
        self$batchnorm4() %>%
        self$flat()
    }
  })
}

```

> The decoder starts from this -- flat -- representation and decompresses it into a time sequence. In both encoder and decoder

Reviewer 2: I think you're using the dashes here for emphasis. You might consider bold or italics instead. Feel free to ignore, as always. 🤗

Author: yeah you know, I think I really want the dashes, not <em></em> so to say ... 
What you said about "visual" before, that might explain it - for me, I hear a spoken sentence, and there's a pause before and after "flat"

> (de-) conv layers, parameters are chosen to handle a sequence length (`num_timesteps`) of 120, which is what we'll use for

Reviewer 2: ```suggestion
(de-)conv layers, parameters are chosen to handle a sequence length (`num_timesteps`) of 120, which is what we'll use for
```

    self$conv2 <- layer_conv_1d_transpose(kernel_size = 11,
                                          filters = 32,
                                          strides = 3)
    self$act2 <- layer_activation_leaky_relu()
    self$batchnorm2 <- layer_batch_normalization()
    self$conv3 <- layer_conv_1d_transpose(
      kernel_size = 9,
      filters = 16,
      strides = 2,
      output_padding = 1
    )
    self$act3 <- layer_activation_leaky_relu()
    self$batchnorm3 <- layer_batch_normalization()
    self$conv4 <- layer_conv_1d_transpose(
      kernel_size = 7,
      filters = 1,
      strides = 1,
      activation = "linear"
    )
    self$batchnorm4 <- layer_batch_normalization()
    
    function (x, mask = NULL) {
      x %>%
        self$reshape() %>%
        self$conv1() %>%
        self$act1() %>%
        self$batchnorm1() %>%
        self$conv2() %>%
        self$act2() %>%
        self$batchnorm2() %>%
        self$conv3() %>%
        self$act3() %>%
        self$batchnorm3() %>%
        self$conv4() %>%
        self$batchnorm4()
    }
  })
}
```

Note that even though we called these constructors `vae_encoder_model()` and `vae_decoder_model()`, there is nothing

> variational to these models per se; they really just are an encoder and a decoder, respectively. Metamorphosis into a VAE will

Reviewer 2: ```suggestion
variational to these models per se; they are really just an encoder and a decoder, respectively. Metamorphosis into a VAE will
```

Reviewer 2: I think this is one of those word order things in English where there's probably no rule, but I'm going on _sprachgefühl_.

Author: top :-)

knitr::include_graphics("images/roessler.png")
```

Like we did for the Lorenz system in the first part of this series, we use `deSolve` to generate data from the Roessler
equations.

```{r, eval = FALSE}
library(deSolve)

parameters <- c(a = .2,
                b = .2,
                c = 5.7)

initial_state <-
  c(x = 1,
    y = 1,
    z = 1.05)

roessler <- function(t, state, parameters) {
  with(as.list(c(state, parameters)), {
    dx <- -y - z
    dy <- x + a * y
    dz = b + z * (x - c)
    
    list(c(dx, dy, dz))
  })
}

times <- seq(0, 2500, length.out = 20000)

roessler_ts <-
  ode(
    y = initial_state,
    times = times,
    func = roessler,
    parms = parameters,
    method = "lsoda"
  ) %>% unclass() %>% as_tibble()

n <- 10000
roessler <- roessler_ts$x[1:n]

roessler <- scale(roessler)
```

Then, noise is added, to the desired degree, by drawing from a normal distribution, centered at zero, with standard deviations
varying between 1 and 2.5.

```{r, eval = FALSE}
# add noise
noise <- 1 # also used 1.5, 2, 2.5
roessler <- roessler + rnorm(10000, mean = 0, sd = noise)

```

> Here you can compare effects of adding no (left), standard deviation-1 (middle), and standard deviation-2.5 Gaussian noise:

Reviewer 2: ```suggestion
Here you can compare effects of adding no standard deviation (left), standard deviation-1 (middle), and standard deviation-2.5 Gaussian noise:
```

Reviewer 2: Or you could change "no" to "nothing"

Author: hm, that was supposed to refer to "noise" an infinity of words later ;-)
could replace by "no noise" I guess

ds_train <- tensor_slices_dataset(list(x_train, y_train)) %>%
  dataset_shuffle(nrow(x_train)) %>%
  dataset_batch(batch_size)

x_test <- test[ , 1:n_timesteps, , drop = FALSE]
y_test <- test[ , (n_timesteps + 1):(2*n_timesteps), , drop = FALSE]

ds_test <- tensor_slices_dataset(list(x_test, y_test)) %>%
  dataset_batch(nrow(x_test))
```

## Results

The LSTM used for comparison with the VAE described above is identical to the architecture employed in the previous post.
While with the VAE, an `fnn_multiplier` of 1 yielded sufficient regularization for all noise levels, some more experimentation
was needed for the LSTM: At noise levels 2 and 2.5, that multiplier was set to 5.

As a result, in all cases, there was one latent variable with high variance and a second one of minor importance. For all
others, variance was close to 0.

*In all cases* here means: In all cases where FNN regularization was used. As already hinted at in the introduction, the main
regularizing factor providing robustness to noise here seems to be FNN loss, not KL divergence. So for all noise levels,
besides FNN-regularized LSTM and VAE models we also tested their non-constrained counterparts.

#### Low noise

> Seeing how all models performed bravourously on the original deterministic series, a noise level of 1 can almost be treated as

Reviewer 2: I don't know the word bravourously. 😳

Author: oops - that happens when you think "yeah that must exist in English as well, as it's from French anyway" and are too lazy to look it up :-(
guess i need sth like "superbly"

> a baseline. Here you see sixteen 120-timestep predictions from both regularized models, FNN-VAE (dark blue) and FNN-LSTM

Reviewer 2: ```suggestion
a baseline. Here you see sixteen 120-timestep predictions from both regularized models, FNN-VAE (dark blue), and FNN-LSTM
```

green, also spanning the whole sequence, we have the original Roessler data, the way they would look had no noise been added.

```{r, eval=TRUE, echo=FALSE, layout="l-page", fig.cap = "Roessler series with added Gaussian noise of standard deviation 1. Grey: actual (noisy) test data. Green: underlying Roessler system. Orange: Predictions from FNN-LSTM. Dark blue: Predictions from FNN-VAE."}
knitr::include_graphics("images/noise_1.png")
```

Despite the noise, forecasts from both models look excellent. Is this due to the FNN regularizer?

Looking at forecasts from their unregularized counterparts, we have to admit these do not look any worse. (For better
comparability, the sixteen sequences to forecast were initiallly picked at random, but used to test all models and
conditions.)

```{r, eval=TRUE, echo=FALSE, layout="l-page", fig.cap = "Roessler series with added Gaussian noise of standard deviation 1. Grey: actual (noisy) test data. Green: underlying Roessler system. Orange: Predictions from unregularized LSTM. Dark blue: Predictions from unregularized VAE."}
knitr::include_graphics("images/noise_1_nofnn.png")
```

What happens when we start to add noise?

#### Substantial noise

> Between noise levels 1.5 and 2, something changed, or became noticeable from visual inspection. Let's directly jump to the

Reviewer 2: ```suggestion
Between noise levels 1.5 and 2, something changed, or became noticeable from visual inspection. Let's jump directly to the
```

highest-used level though: 2.5.

Here first are predictions obtained from the unregularized models.

```{r, eval=TRUE, echo=FALSE, layout="l-page", fig.cap = "Roessler series with added Gaussian noise of standard deviation 2.5. Grey: actual (noisy) test data. Green: underlying Roessler system. Orange: Predictions from unregularized LSTM. Dark blue: Predictions from unregularized VAE."}
knitr::include_graphics("images/noise_2.5_nofnn.png")
```

> Both LSTM and VAE, but the latter to an even higher degree, get "distracted" a bit too much by the noise, leading to cases

Reviewer 2: Here's another one where you might move the caveat to _after_ the main clause, e.g., _Both LSTM and VAE get "distracted" a bit too much by the noise, the latter to an even higher degree. This leads to cases where…_

Author: interesting - same thing as always, that "focus" thing which probably is due to some exaggerated wish to control attention ..

where predictions strongly "overshoot" the underlying non-noisy rhythm. This is not surprising, of course: They were *trained*
on the noisy version; predict fluctuations is what they learned.

Do we see the same with the FNN models?

```{r, eval=TRUE, echo=FALSE, layout="l-page", fig.cap = "Roessler series with added Gaussian noise of standard deviation 2.5. Grey: actual (noisy) test data. Green: underlying Roessler system. Orange: Predictions from FNN-LSTM. Dark blue: Predictions from FNN-VAE."}
knitr::include_graphics("images/noise_2.5.png")
```

Interestingly, we see a much better fit to the underlying Roessler system now! Especially the VAE model, FNN-VAE, surprises
with a whole new smoothness of predictions; but FNN-LSTM turns up much smoother forecasts as well.

"Smooth, fitting the system..." -- by now you may be wondering, when are we going to come up with more quantitative
assertions? If quantitative implies "mean squared error" (MSE), and if MSE is taken to be some divergence between forecasts
and the true target from the test set, the answer is that this MSE doesn't differ much between any of the four architectures.

> Put differently, it mostly is a function of noise level.

Reviewer 2: ```suggestion
Put differently, it is mostly a function of noise level.
```
